src/controller.rs

```diff
@@ -125,7 +131,25 @@ async fn reconcile(app_instance: Arc<AppInstance>, ctx: Arc<Context>) -> Result<
         return Ok(Action::await_change());
     }
 
-    let state = reconciliation_state(&app_instance, &ctx).await?;
+    let app_instance_api: Api<AppInstance> =
+        Api::namespaced(ctx.client.clone(), &app_instance.namespace_any());
+    finalizer(
+        &app_instance_api,
+        KUBIT_FINALIZER,
+        app_instance,
+        |event| async {
+            match event {
+                Finalizer::Apply(app_instance) => reconcile_apply(&app_instance, &ctx).await,
+                Finalizer::Cleanup(app_instance) => reconcile_cleanup(&app_instance, &ctx).await,
+            }
```

Comment:
Introducing a finalizer causes a split in the flow of the controller between these 2 events. The `Apply` event is what we had before, but as a single flow since there was no cleanup/deletion process. The `Cleanup` `match` arm is entirely new with its logic implemented in this PR.

I think it works nicely as is, but if there are any objections I can change this to something else which isn't using this `reconcile_*` style.

I used [`kube-rs/controller-rs`](https://github.com/kube-rs/controller-rs/blob/6adbb244c5f35e3d84b4040f824d088e8c9e6bfd/src/controller.rs#L69-L76) as a reference implementation for this.